Developer Features: Add Tests for Analytics Tracking

Aurorum opened the pull request:
Related to #87474

## Proposed Changes

This simplifies/fixes the analytics tracking for Developer Features. I couldn't actually reproduce the incorrect data, but @wongasy reported it here: https://github.com/Automattic/wp-calypso/pull/87474#issuecomment-1958461911.

~~Originally, this wasn't hardcoded because there was no need for an `id` - that was added in the last minute in order to avoid errors when mapping the features list. As the only condition for the `id` is that it's unique, I see no reason why we can't save a bit of code and use it for recording analytics too. That also guarantees no wrong analytics if the link structure changes in future.~~

As per discussion below, this now just adds tests for the existing approach rather than changing it. This ensures that analytics won't break in future.

## Testing Instructions

Run `yarn test-client client/me/developer/features/test/use-handle-click-link.js` and confirm that they pass. You can change the feature slug in the tests locally to confirm that they fail when incorrect.

cc @nightnei, @oswian, @dlind1

Aurorum committed:
Create branch

Aurorum committed:
Developer Features: Simplify/Fix Analytics Tracking

nightnei commented:
@Aurorum of course it's my subjective opinion, but I see some disadvantages with such approach:

1) The main my point - with the current approach, we have consistency - we have a separate handler that always works the same way with links - takes the `href` and handles it to provide the `feature` property. And with proposed approach we don't have consistency - sometimes we use `id`, sometimes we use hardcoded string and sometimes (for example the last one item) we have `id` but it's not used for analytics, since we don't have `Learn more` link.
2) Less critical and mostly subjective - with the current approach, we have a separate handler where all logic happens, so the actual components look clearer - there is only the handler name, so it's much easier to read such components.
3) Also, using `href` was discussed and it was agreed, so now it can mess the analytics. So at least we should keep the same `feature` names.

In general, sometimes less code looks harder to read, and readability is much more critical than number of lines, but of course it's subjective opinion and it depend on the specific case. In this case, I would keep what we have.

@wongasy WDYT?

wongasy commented:
>I couldn't actually reproduce the incorrect data, but @wongasy reported it here: https://github.com/Automattic/wp-calypso/pull/87474#issuecomment-1958461911.

I suspect the faulty data I saw were sent before this commit https://github.com/Automattic/wp-calypso/pull/87474/commits/9fba17ba1bb4f1ad282f6be44f33cd627d796be3. Looks like we are good in production.

wongasy commented:
>In general, sometimes less code looks harder to read, and readability is much more critical than number of lines

I agree. Although, in this case, I do find the changes proposed in this PR easier to read. Namely, we don't have to bother with `useCallback`, nor do we have to worry about extra logic to strip the feature name from the URL. But like you said, it could be a personal preference. 

>with the current approach, we have consistency - we have a separate handler that always works the same way with links - takes the `href` and handles it to provide the `feature` property. 

On the other hand, I also take the point that, with the new approach, we end up with two click handlers. If our tracking logic changes, it's easy to forget to update _both_ handlers.

Given that we've confirmed that nothing is broken. Let's keep things as-is. We'll refactor as needed when we encounter scenarios that require a more robust mechanism for tracking. For now, the priority is to ensure we're tracking the correct data, which has been confirmed!

Aurorum committed:
Add unit tests instead

Aurorum commented:
Yes, I think it's down to personal preference - that reasoning makes sense. :) If we're definitely set on this approach though, I think it's worth adding unit tests so that this doesn't happen again, as per https://github.com/Automattic/wp-calypso/pull/87474#pullrequestreview-1883998182. That at least ensures the analytics data doesn't change in future, since it can be easy to miss for someone not familiar with this section!

What do you both think of this PR now? :)

Aurorum committed:
Correct version

Aurorum commented:
Sorry, just noticed I committed the wrong version 🤦 This one should be good to review, and would've caught the issue.

<img width="775" alt="Screenshot 2024-02-25 at 13 26 10" src="https://github.com/Automattic/wp-calypso/assets/43215253/4185c04e-5b66-462b-8527-7cb0b2357da7">

cc @oswian if that's what you had in mind. :)

Aurorum committed:
Remove conditional expect

oswian commented:
> cc @oswian if that's what you had in mind. :)

Thanks for working on the tests @Aurorum.

I have a few suggestions:

- Avoid re-implementing the `useHandleClickLink` logic inside the test itself. We want to test the behaviour of the hook without making assumptions about the underlying implementation.
- Remove the call to `useFeaturesList`. The list of features may change over time, and while we _could_ list all of the support URL's we know about today, I would approach writing the test in a more generic way ie. focus on how it handles the 'shape' of different URL's.
- Clarify and be specific regarding the intent of the test in the description and by providing clear inputs and outputs

Below is a test which I think can replace the two you proposed. The URL's I've included are variations that reflect the shape of the support URL's I'm aware of, but feel free to add/change.

```ts
describe( 'useHandleClickLink', () => {
	...

	describe( 'tracks event is recorded without the /support/ prefix', () => {
		test.each( [
			[ 'https://wordpress.com/support/feature', 'feature' ],
			[ 'https://wordpress.com/support/category/feature', 'category/feature' ],
			[ 'https://wordpress.com/support/feature-with-hyphens', 'feature-with-hyphens' ],
			[ 'https://wordpress.com/support/supported/feature', 'supported/feature' ],
		] )(
			"for the URL %s a tracks event is recorded using the 'feature' value '%s'",
			( href, feature ) => {
				const { result } = renderHook( () => useHandleClickLink() );
				const event = { currentTarget: { href } };

				act( () => result.current( event ) );

				expect( mockRecordTracksEventWithUserIsDevAccount ).toHaveBeenCalledWith( tracksHandle, {
					feature,
				} );
			}
		);
	} );
} );
```

Test run output:

```
 PASS  client/me/developer/features/test/use-handle-click-link.js
  useHandleClickLink
    tracks event is recorded without the /support/ prefix
      ✓ for the URL https://wordpress.com/support/feature a tracks event is recorded using the 'feature' value 'feature' (9 ms)
      ✓ for the URL https://wordpress.com/support/category/feature a tracks event is recorded using the 'feature' value 'category/feature' (1 ms)
      ✓ for the URL https://wordpress.com/support/feature-with-hyphens a tracks event is recorded using the 'feature' value 'feature-with-hyphens' (1 ms)
      ✓ for the URL https://wordpress.com/support/supported/feature a tracks event is recorded using the 'feature' value 'supported/feature' (1 ms)
```

Aurorum committed:
Address review

Aurorum commented:
Thanks @oswian! Totally agree a more generic test is better, and I've added your recommended proposal now.

> Avoid re-implementing the useHandleClickLink logic inside the test itself. We want to test the behaviour of the hook without making assumptions about the underlying implementation.

I thought it might be helpful if I clarified that my intention for that test was to prevent a repeat of #87474, which is why I tested the links in `useFeaturesList`. The idea is that it would prevent someone adding a link with the wrong structure that would break the analytics in future. 

I've removed it now, but based on that, let me know if you think it'd still make sense to include. :)

wongasy commented on client/me/developer/features/test/use-handle-click-link.js:
```diff
@@ -0,0 +1,44 @@
+		);
+	} );
+
+	describe( 'ensure track events are recorded without the /support/ prefix', () => {
+		test.each( [
+			[ 'https://wordpress.com/support/feature', 'feature' ],
+			[ 'https://wordpress.com/support/category/feature', 'category/feature' ],
+			[ 'https://wordpress.com/support/feature-with-hyphens', 'feature-with-hyphens' ],
+			[ 'https://wordpress.com/support/supported/feature', 'supported/feature' ],
+		] )(
```
Thanks @Aurorum for adding unit tests to validate analytics values! 

>The idea is that it would prevent someone adding a link with the wrong structure that would break the analytics in future.

I'm not sure if we should restrict the structure of the link. We should allow the flexibility to supply any link. If the current analytics logic doesn't handle that link, then we would adjust the analytics logic accordingly when we cross the bridge, while maintaining the existing values that have been used to track events.

At the end of the day, what we need to ensure is that existing tracking values remain in tact when changes are made. So I would go as simple as listing out all the links we have and their expected tracking value, instead of trying to check for the removal of `/support` prefix. 

```suggestion
	describe( 'ensure track events are recorded with the expected feature name', () => {
		test.each( [
			[ 'https://wordpress.com/support/connect-to-ssh-on-wordpress-com', 'connect-to-ssh-on-wordpress-com', ],
			[ 'https://wordpress.com/support/how-to-create-a-staging-site/', 'how-to-create-a-staging-site', ],
		],
			[ 'https://wordpress.com/support/code', 'code', ],
			[ 'https://wordpress.com/support/domains/https-ssl', 'domains/https-ssl', ],
			[ 'https://wordpress.com/support/help-support-options', 'help-support-options', ],
			[ 'https://wordpress.com/support/restore', 'restore', ],
			[ 'https://wordpress.com/support/malware-and-site-security', 'malware-and-site-security', ],
			[ 'https://wordpress.com/support/site-monitoring', 'site-monitoring', ],
		] )(
```

oswian replied:
Hi @wongasy, nice to meet you 👋 

The two unit tests I've proposed are for `useHandleClickLink`. This hook, by design, has no knowledge of the specific links on the `/me/developer` page. It does however contain explicit handling for the `/support/` prefix and trailing `/`, and so it's this logic that the unit tests cover. ie.

The test "ensure track events are recorded without the /support/ prefix" covers these lines:

https://github.com/Automattic/wp-calypso/blob/582dcb267934545211826b17e1267d3b7f737402/client/me/developer/features/use-handle-click-link.tsx#L9-L15

While the test "ensure track events are recorded without a trailing /" covers this line:

https://github.com/Automattic/wp-calypso/blob/582dcb267934545211826b17e1267d3b7f737402/client/me/developer/features/use-handle-click-link.tsx#L16

To verify the tracking values for the _actual_ links on the `/me/developer` page, I would suggest creating a test for the `DeveloperFeatures` component. This component has knowledge of the specific links that we want to validate the tracking values for (via it's internal call to `useFeaturesList()`). The test would roughly take the form:

- Render the `DeveloperFeatures` component (using the same approach to mock `useRecordTracksEventWithUserIsDevAccount`)
- Click each link eg. `https://wordpress.com/support/connect-to-ssh-on-wordpress-com`
- Verify the tracking values

wongasy replied:
Thanks @oswian for the explainer. I see your point that, it makes better sense to test the actual links in relation to the component that defines the links, instead of the helper function which doesn't have knowledge of the links.

wongasy reviewed (approved):
I've ran the build all unit tests passed 🎉 I'll go ahead and deploy this.
